File: test_e2e.py
"""
E2E integration test for FinSight backend.
Tests all major endpoints in a realistic user flow.
Run: python test_e2e.py
Requires: backend running on localhost:8000, MySQL finsight database
"""
import urllib.request, urllib.error, json, sys, time

BASE = "http://localhost:8000"
passed = 0
failed = 0

def req(method, path, body=None, token=None):
    url = f"{BASE}{path}"
    headers = {"Content-Type": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    data = json.dumps(body).encode() if body else None
    r = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        resp = urllib.request.urlopen(r, timeout=10)
        return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        return {"_error": e.code, "_detail": e.read().decode()}

def check(name, ok, detail=""):
    global passed, failed
    if ok:
        passed += 1
        print(f"  PASS  {name}")
    else:
        failed += 1
        print(f"  FAIL  {name}  {detail}")

# ── 1. Health ────────────────────────────────────────────
print("\n=== 1. Health Check ===")
h = req("GET", "/health")
check("health endpoint", h.get("status") == "healthy", str(h))

# ── 2. Register ──────────────────────────────────────────
print("\n=== 2. Register ===")
import random
suffix = random.randint(10000, 99999)
email = f"testuser{suffix}@example.com"
register_body = {"full_name": "Test User", "email": email, "password": "testpass123"}
r = req("POST", "/auth/register", register_body)
token = r.get("access_token", "")
check("register returns token", bool(token), str(r))

# ── 3. Login ─────────────────────────────────────────────
print("\n=== 3. Login ===")
r = req("POST", "/auth/login", {"email": email, "password": "testpass123"})
token = r.get("access_token", "")
check("login returns token", bool(token), str(r))
user_id = r.get("user_id", 0)
check("login returns user_id", user_id > 0)

# ── 4. Auth/Me ───────────────────────────────────────────
print("\n=== 4. Get Profile ===")
r = req("GET", "/auth/me", token=token)
check("auth/me returns email", r.get("email") == email, str(r))
check("auth/me has currency", bool(r.get("currency")))

# ── 5. Categories ────────────────────────────────────────
print("\n=== 5. Categories ===")
r = req("GET", "/categories", token=token)
check("categories returns list", isinstance(r, list), str(r))
check("categories has items", len(r) >= 8)
income_cats = [c for c in r if c["type"] == "income"]
expense_cats = [c for c in r if c["type"] == "expense"]
check("has income categories", len(income_cats) >= 3)
check("has expense categories", len(expense_cats) >= 5)

# ── 6. Create Transaction ────────────────────────────────
print("\n=== 6. Transactions ===")
cat_id = expense_cats[0]["id"]
r = req("POST", "/transactions", {
    "category_id": cat_id, "amount": 5000, "type": "expense",
    "description": "Test expense", "transaction_date": "2026-07-01"
}, token=token)
tx_id = r.get("id", 0)
check("create transaction returns id", tx_id > 0, str(r))

# Income
inc_cat_id = income_cats[0]["id"]
r = req("POST", "/transactions", {
    "category_id": inc_cat_id, "amount": 100000, "type": "income",
    "description": "Test income", "transaction_date": "2026-07-01"
}, token=token)
inc_tx_id = r.get("id", 0)
check("create income transaction", inc_tx_id > 0)

# ── 7. List Transactions ─────────────────────────────────
print("\n=== 7. List Transactions ===")
r = req("GET", "/transactions", token=token)
check("list transactions", r.get("total", 0) >= 2, str(r))

# ── 8. Soft Delete Transaction ───────────────────────────
print("\n=== 8. Soft Delete ===")
r = req("DELETE", f"/transactions/{tx_id}?soft=true", token=token)
check("soft delete", "deleted" in str(r).lower(), str(r))

# ── 9. Restore Transaction ───────────────────────────────
print("\n=== 9. Restore ===")
r = req("POST", f"/transactions/{tx_id}/restore", token=token)
check("restore", "restored" in str(r).lower(), str(r))

# ── 10. Goals ────────────────────────────────────────────
print("\n=== 10. Goals ===")
r = req("POST", "/goals", {
    "name": "Test Goal", "target_amount": 50000, "deadline": "2027-12-31"
}, token=token)
goal_id = r.get("id", 0)
check("create goal returns id", goal_id > 0, str(r))

# List goals
r = req("GET", "/goals", token=token)
check("list goals", len(r) >= 1, str(r))

# Fund goal
r = req("POST", f"/goals/{goal_id}/fund", {"amount": 10000}, token=token)
check("fund goal", "added" in str(r).lower(), str(r))

# Complete goal
r = req("POST", f"/goals/{goal_id}/complete", token=token)
check("complete goal", "completed" in str(r).lower(), str(r))

# Delete goal
r = req("DELETE", f"/goals/{goal_id}", token=token)
check("delete goal", "deleted" in str(r).lower(), str(r))

# Verify deletion
r = req("GET", "/goals", token=token)
check("goal deleted from list", all(g["id"] != goal_id for g in r))

# ── 11. Budgets ──────────────────────────────────────────
print("\n=== 11. Budgets ===")
r = req("POST", "/budgets", {"category_id": cat_id, "monthly_limit": 25000}, token=token)
check("create budget", "set" in str(r).lower(), str(r))

# Get utilization
import urllib.parse
r = req("GET", f"/budgets/utilization?month=7&year=2026", token=token)
check("budget utilization", isinstance(r, (list, dict)), str(r))
if isinstance(r, (list, dict)):
    items = r if isinstance(r, list) else list(r.values())
    check("utilization has entries", len(items) >= 1)

# ── 12. Dashboard ────────────────────────────────────────
print("\n=== 12. Dashboard ===")
r = req("GET", "/dashboard?month=7&year=2026", token=token)
check("dashboard returns data", bool(r.get("total_income") is not None), str(r))
check("dashboard has income", r.get("total_income", 0) > 0)
check("dashboard has expense", r.get("total_expense", 0) > 0)
check("dashboard has net_savings", r.get("net_savings") is not None)
check("dashboard has monthly_trends", len(r.get("monthly_trends", [])) >= 1)
check("dashboard has top_categories", len(r.get("top_categories", [])) >= 1)

# ── 13. Analytics Trends ─────────────────────────────────
print("\n=== 13. Analytics ===")
r = req("GET", "/analytics/trends?months=12", token=token)
check("trends returns list", isinstance(r, list), str(r))
if isinstance(r, list) and len(r) > 0:
    check("trends has month field", "month" in r[0])
    check("trends has income field", "income" in r[0])

# ── 14. Insights: Predict ────────────────────────────────
print("\n=== 14. ML: Predict ===")
r = req("GET", "/insights/predict", token=token)
check("predict returns data", bool(r), str(r))
check("predict has predicted_total", r.get("predicted_total") is not None)

# ── 15. Insights: Suggest Category ───────────────────────
print("\n=== 15. ML: Suggest Category ===")
desc = urllib.parse.quote("uber ride to airport")
r = req("GET", f"/insights/suggest-category?description={desc}", token=token)
check("suggest returns category", r.get("category") is not None, str(r))
check("suggest has category_id", r.get("category_id") is not None)

# ── 16. Insights: Cluster ────────────────────────────────
print("\n=== 16. ML: Cluster ===")
r = req("GET", "/insights/cluster", token=token)
check("cluster returns list", isinstance(r, list), str(r))
if isinstance(r, list) and len(r) > 0:
    check("cluster has cluster_id", "cluster_id" in r[0])
    check("cluster has name", "name" in r[0])

# ── 17. Currency Rates ──────────────────────────────────
print("\n=== 17. Currency ===")
r = req("GET", "/currency/rates", token=token)
check("rates returns data", bool(r), str(r))
check("rates has base", r.get("base") == "INR")
check("rates has INR rate", "INR" in r.get("rates", {}))

# ── 18. Report ──────────────────────────────────────────
print("\n=== 18. Report ===")
r = req("POST", "/report/generate", token=token)
check("report generates", r.get("path") is not None, str(r))

# ── 19. Error Handling ──────────────────────────────────
print("\n=== 19. Error Handling ===")
# Invalid login
r = req("POST", "/auth/login", {"email": "nonexistent@x.com", "password": "wrong"})
check("invalid login returns error", "_error" in r and r["_error"] == 401, str(r))

# Unauthorized access
r = req("GET", "/transactions")
check("no-token returns 401", "_error" in r and r["_error"] == 401, str(r))

# ── Summary ──────────────────────────────────────────────
print(f"\n{'='*50}")
print(f"RESULTS:  {passed} passed, {failed} failed out of {passed+failed} tests")
if failed == 0:
    print("ALL TESTS PASSED")
else:
    print(f"SOME TESTS FAILED")
    sys.exit(1)

File: test_e2e.js
/**
 * E2E integration test for FinSight backend in Node.js.
 * Tests all major endpoints in a realistic user flow.
 * Run: node test_e2e.js
 * Requires: backend running on localhost:8000, MySQL finsight database
 */
const http = require('http');

const BASE_URL = 'http://localhost:8000';
let passed = 0;
let failed = 0;

function req(method, path, body = null, token = null) {
  return new Promise((resolve) => {
    const parsedUrl = new URL(BASE_URL + path);
    const options = {
      hostname: parsedUrl.hostname,
      port: parsedUrl.port || 80,
      path: parsedUrl.pathname + parsedUrl.search,
      method: method,
      headers: {
        'Content-Type': 'application/json'
      }
    };
    if (token) {
      options.headers['Authorization'] = `Bearer ${token}`;
    }

    let postData = null;
    if (body) {
      postData = JSON.stringify(body);
      options.headers['Content-Length'] = Buffer.byteLength(postData);
    }

    const reqObj = http.request(options, (res) => {
      let data = [];
      res.on('data', (chunk) => data.push(chunk));
      res.on('end', () => {
        const raw = Buffer.concat(data).toString();
        if (res.statusCode >= 400) {
          resolve({
            _error: res.statusCode,
            _detail: raw
          });
        } else {
          try {
            resolve(JSON.parse(raw));
          } catch (e) {
            resolve(raw);
          }
        }
      });
    });

    reqObj.on('error', (err) => {
      resolve({
        _error: 500,
        _detail: err.message
      });
    });

    if (postData) {
      reqObj.write(postData);
    }
    reqObj.end();
  });
}

function check(name, ok, detail = "") {
  if (ok) {
    passed++;
    console.log(`  PASS  ${name}`);
  } else {
    failed++;
    console.log(`  FAIL  ${name}  ${detail}`);
  }
}

async function run() {
  try {
    // ── 1. Health ────────────────────────────────────────────
    console.log("\n=== 1. Health Check ===");
    const h = await req("GET", "/health");
    check("health endpoint", h.status === "healthy", JSON.stringify(h));

    // ── 2. Register ──────────────────────────────────────────
    console.log("\n=== 2. Register ===");
    const suffix = Math.floor(10000 + Math.random() * 90000);
    const email = `testuser${suffix}@example.com`;
    const registerBody = { full_name: "Test User", email: email, password: "testpass123" };
    const rRegister = await req("POST", "/auth/register", registerBody);
    const token = rRegister.access_token || "";
    check("register returns token", !!token, JSON.stringify(rRegister));

    // ── 3. Login ─────────────────────────────────────────────
    console.log("\n=== 3. Login ===");
    const rLogin = await req("POST", "/auth/login", { email: email, password: "testpass123" });
    const tokenLogin = rLogin.access_token || "";
    check("login returns token", !!tokenLogin, JSON.stringify(rLogin));
    const userId = rLogin.user_id || 0;
    check("login returns user_id", userId > 0, JSON.stringify(rLogin));

    const activeToken = tokenLogin || token;

    // ── 4. Auth/Me ───────────────────────────────────────────
    console.log("\n=== 4. Get Profile ===");
    const rMe = await req("GET", "/auth/me", null, activeToken);
    check("auth/me returns email", rMe.email === email, JSON.stringify(rMe));
    check("auth/me has currency", !!rMe.currency, JSON.stringify(rMe));

    // ── 5. Categories ────────────────────────────────────────
    console.log("\n=== 5. Categories ===");
    const rCats = await req("GET", "/categories", null, activeToken);
    check("categories returns list", Array.isArray(rCats), JSON.stringify(rCats));
    check("categories has items", Array.isArray(rCats) && rCats.length >= 8);
    const incomeCats = Array.isArray(rCats) ? rCats.filter(c => c.type === "income") : [];
    const expenseCats = Array.isArray(rCats) ? rCats.filter(c => c.type === "expense") : [];
    check("has income categories", incomeCats.length >= 3);
    check("has expense categories", expenseCats.length >= 5);

    // ── 6. Create Transaction ────────────────────────────────
    console.log("\n=== 6. Transactions ===");
    const catId = expenseCats[0] ? expenseCats[0].id : 1;
    const rTx = await req("POST", "/transactions", {
      category_id: catId, amount: 5000, type: "expense",
      description: "Test expense", transaction_date: "2026-07-01"
    }, activeToken);
    const txId = rTx.id || 0;
    check("create transaction returns id", txId > 0, JSON.stringify(rTx));

    // Income
    const incCatId = incomeCats[0] ? incomeCats[0].id : 2;
    const rIncTx = await req("POST", "/transactions", {
      category_id: incCatId, amount: 100000, type: "income",
      description: "Test income", transaction_date: "2026-07-01"
    }, activeToken);
    const incTxId = rIncTx.id || 0;
    check("create income transaction", incTxId > 0, JSON.stringify(rIncTx));

    // ── 7. List Transactions ─────────────────────────────────
    console.log("\n=== 7. List Transactions ===");
    const rList = await req("GET", "/transactions", null, activeToken);
    check("list transactions", rList.total >= 2, JSON.stringify(rList));

    // ── 8. Soft Delete Transaction ───────────────────────────
    console.log("\n=== 8. Soft Delete ===");
    const rDel = await req("DELETE", `/transactions/${txId}?soft=true`, null, activeToken);
    check("soft delete", JSON.stringify(rDel).toLowerCase().includes("deleted"), JSON.stringify(rDel));

    // ── 9. Restore Transaction ───────────────────────────────
    console.log("\n=== 9. Restore ===");
    const rRest = await req("POST", `/transactions/${txId}/restore`, null, activeToken);
    check("restore", JSON.stringify(rRest).toLowerCase().includes("restored"), JSON.stringify(rRest));

    // ── 10. Goals ────────────────────────────────────────────
    console.log("\n=== 10. Goals ===");
    const rGoal = await req("POST", "/goals", {
      name: "Test Goal", target_amount: 50000, deadline: "2027-12-31"
    }, activeToken);
    const goalId = rGoal.id || 0;
    check("create goal returns id", goalId > 0, JSON.stringify(rGoal));

    // List goals
    const rListGoals = await req("GET", "/goals", null, activeToken);
    check("list goals", Array.isArray(rListGoals) && rListGoals.length >= 1, JSON.stringify(rListGoals));

    // Fund goal
    const rFund = await req("POST", `/goals/${goalId}/fund`, { amount: 10000 }, activeToken);
    check("fund goal", JSON.stringify(rFund).toLowerCase().includes("added"), JSON.stringify(rFund));

    // Complete goal
    const rComp = await req("POST", `/goals/${goalId}/complete`, null, activeToken);
    check("complete goal", JSON.stringify(rComp).toLowerCase().includes("completed"), JSON.stringify(rComp));

    // Delete goal
    const rDelGoal = await req("DELETE", `/goals/${goalId}`, null, activeToken);
    check("delete goal", JSON.stringify(rDelGoal).toLowerCase().includes("deleted"), JSON.stringify(rDelGoal));

    // Verify deletion
    const rVerifyGoal = await req("GET", "/goals", null, activeToken);
    check("goal deleted from list", Array.isArray(rVerifyGoal) && rVerifyGoal.every(g => g.id !== goalId), JSON.stringify(rVerifyGoal));

    // ── 11. Budgets ──────────────────────────────────────────
    console.log("\n=== 11. Budgets ===");
    const rBudget = await req("POST", "/budgets", { category_id: catId, monthly_limit: 25000 }, activeToken);
    check("create budget", JSON.stringify(rBudget).toLowerCase().includes("set"), JSON.stringify(rBudget));

    // Get utilization
    const rUtil = await req("GET", "/budgets/utilization?month=7&year=2026", null, activeToken);
    check("budget utilization", typeof rUtil === 'object', JSON.stringify(rUtil));
    const items = Array.isArray(rUtil) ? rUtil : Object.values(rUtil || {});
    check("utilization has entries", items.length >= 1, JSON.stringify(items));

    // ── 12. Dashboard ────────────────────────────────────────
    console.log("\n=== 12. Dashboard ===");
    const rDash = await req("GET", "/dashboard?month=7&year=2026", null, activeToken);
    check("dashboard returns data", rDash.total_income !== undefined, JSON.stringify(rDash));
    check("dashboard has income", rDash.total_income > 0, JSON.stringify(rDash));
    check("dashboard has expense", rDash.total_expense > 0, JSON.stringify(rDash));
    check("dashboard has net_savings", rDash.net_savings !== undefined, JSON.stringify(rDash));
    check("dashboard has monthly_trends", Array.isArray(rDash.monthly_trends) && rDash.monthly_trends.length >= 1, JSON.stringify(rDash.monthly_trends));
    check("dashboard has top_categories", Array.isArray(rDash.top_categories) && rDash.top_categories.length >= 1, JSON.stringify(rDash.top_categories));

    // ── 13. Analytics Trends ─────────────────────────────────
    console.log("\n=== 13. Analytics ===");
    const rAnal = await req("GET", "/analytics/trends?months=12", null, activeToken);
    check("trends returns list", Array.isArray(rAnal), JSON.stringify(rAnal));
    if (Array.isArray(rAnal) && rAnal.length > 0) {
      check("trends has month field", rAnal[0].month !== undefined);
      check("trends has income field", rAnal[0].income !== undefined);
    } else {
      check("trends has month field", false, "Empty list");
      check("trends has income field", false, "Empty list");
    }

    // ── 14. Insights: Predict ────────────────────────────────
    console.log("\n=== 14. ML: Predict ===");
    const rPred = await req("GET", "/insights/predict", null, activeToken);
    check("predict returns data", !!rPred, JSON.stringify(rPred));
    check("predict has predicted_total", rPred.predicted_total !== undefined || rPred.next_month_prediction !== undefined, JSON.stringify(rPred));

    // ── 15. Insights: Suggest Category ───────────────────────
    console.log("\n=== 15. ML: Suggest Category ===");
    const desc = encodeURIComponent("uber ride to airport");
    const rSugg = await req("GET", `/insights/suggest-category?description=${desc}`, null, activeToken);
    check("suggest returns category", rSugg.category !== undefined, JSON.stringify(rSugg));
    check("suggest has category_id", rSugg.category_id !== undefined, JSON.stringify(rSugg));

    // ── 16. Insights: Cluster ────────────────────────────────
    console.log("\n=== 16. ML: Cluster ===");
    const rClust = await req("GET", "/insights/cluster", null, activeToken);
    check("cluster returns list", Array.isArray(rClust), JSON.stringify(rClust));
    if (Array.isArray(rClust) && rClust.length > 0) {
      check("cluster has cluster_id", rClust[0].cluster_id !== undefined);
      check("cluster has name", rClust[0].name !== undefined);
    } else {
      check("cluster has cluster_id", true, "Empty list (valid if not enough data)");
      check("cluster has name", true, "Empty list (valid if not enough data)");
    }

    // ── 17. Currency Rates ──────────────────────────────────
    console.log("\n=== 17. Currency ===");
    const rCurr = await req("GET", "/currency/rates", null, activeToken);
    check("rates returns data", !!rCurr, JSON.stringify(rCurr));
    check("rates has base", rCurr.base === "INR", JSON.stringify(rCurr));
    check("rates has INR rate", rCurr.rates && rCurr.rates.INR !== undefined, JSON.stringify(rCurr));

    // ── 18. Report ──────────────────────────────────────────
    console.log("\n=== 18. Report ===");
    const rRep = await req("POST", "/report/generate", null, activeToken);
    check("report generates", rRep.path !== undefined, JSON.stringify(rRep));

    // ── 19. Error Handling ──────────────────────────────────
    console.log("\n=== 19. Error Handling ===");
    // Invalid login
    const rBadLogin = await req("POST", "/auth/login", { email: "nonexistent@x.com", password: "wrong" });
    check("invalid login returns error", rBadLogin._error === 401, JSON.stringify(rBadLogin));

    // Unauthorized access
    const rNoToken = await req("GET", "/transactions");
    check("no-token returns 401", rNoToken._error === 401, JSON.stringify(rNoToken));

    // ── Summary ──────────────────────────────────────────────
    console.log(`\n${'='.repeat(50)}`);
    console.log(`RESULTS:  ${passed} passed, ${failed} failed out of ${passed + failed} tests`);
    if (failed === 0) {
      console.log("ALL TESTS PASSED");
      process.exit(0);
    } else {
      console.log("SOME TESTS FAILED");
      process.exit(1);
    }
  } catch (error) {
    console.error("Test execution exception:", error);
    process.exit(1);
  }
}

run();
